Discard null values in complex objects in strategic patch

ardaguclu opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

In strategic patch, if key does not exist in original, value of this
new key is not distilled by discarding the null values.

That brings about ,in key **creation** step, null values are also patched. However,
during the **update** process of this key in subsequent patches, these null values
are deleted and this hurts the idempotency of strategic patch.

```
$ cluster/kubectl.sh create -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
    - name: nginx
      image: nginx
      ports:
        - containerPort: 80
EOF

$ cluster/kubectl.sh patch pod test -p '{"metadata": {"annotations": {"key1": "value1","key2":null}}}'
pod/test patched

$ cluster/kubectl.sh get pod test -o json | jq -r .metadata.annotations
{
  "key1": "value1",
  "key2": ""
}

$ cluster/kubectl.sh patch pod test -p '{"metadata": {"annotations": {"key1": "value1","key2":null}}}'
pod/test patched

$ cluster/kubectl.sh get pod test -o json | jq -r .metadata.annotations
{
  "key1": "value1"
}

```

This PR adds discard mechanism for null values if key does not exist
in original. It traverses all nested objects.

#### Which issue(s) this PR fixes:
Fixes #105689

#### Special notes for your reviewer:
this PR also discards `[]` for slice types accepting as null. Is this relevant or `[]` should not be discarded? 

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ardaguclu committed:
Discard null values in complex objects in strategic patch

In strategic patch, if key does not exist in original, value of this
new key is not distilled by discarding the null values.

That brings about in key creation step, null values are also patched. However,
during the update process of this key in subsequent patches, these null values
are deleted and this hurts the idempotency of strategic patch.

This PR adds discard mechanism for null values if key does not exist
in original. It traverses all nested objects.

ardaguclu commented:
cc: @liggitt

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -6713,6 +6713,36 @@ func TestUnknownField(t *testing.T) {
+			Modified: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{"key":null},"name":"foo","scalar":true}`,
+
+			ExpectedTwoWay:         `{"complex_nullable":{"key":null}}`,
+			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
```
why is `complex_nullable` removed here?

ardaguclu replied:
This is actually the real issue(ref: https://github.com/kubernetes/kubernetes/issues/105689). Because in master branch, after first patch request, `complex_nullable` is preserved but in second patch request(with same commands) it will be removed.

apelisse replied:
Here's what I'm reading in #105689:

Case 1: (working fine)
> If the object's .metadata.labels contains anything other than EFRIED, this works fine: the resulting object has no EFRIED key.
```
Before: {"metadata": {"labels": {"Random": "value"}}}
Patch: {"metadata": {"labels": {"EFRIED": null}}}
After: {"metadata": {"labels": {"Random": "value"}}}
```

Case 2: (broken case, I can't see a test for that?)
> However, if .metadata.labels is initially absent, this command adds it with an EFRIED entry with the empty string as a value
```
Before: {"metadata": {}}
Patch: {"metadata": {"labels": {"EFRIED": null}}}
After: {"metadata": {"labels": {"EFRIED": ""}}}
I would expect (either, I'm not 100% sure, I think both would be fine from OPs perspective): 
{"metadata": {"labels": {}}
OR
{"metadata": {}}
```

Case 3: (working fine?)
> Running the command again removes the entry and the labels.
```
Before: {"metadata": {"labels": {"EFRIED": null}}}
Patch: {"metadata": {"labels": {"EFRIED": null}}}
After: {"metadata": {"labels": {}}
I would expect: {"metadata": {"labels": {}}}
```
I think OP is agreeing with you that `labels` is removed, I think that's a bug. For these situations, I think it should behave exactly like json merge patch does (I'd love to be proven wrong though, I can't seem to verify how json merge patch would behave right now).

ardaguclu replied:
Case 1 and Case 3 are working as expected. Case 2 is failing.

Case 2 is failing because this [line](https://github.com/kubernetes/kubernetes/blob/876d4e0ab029ac7c314bb0e033bdd036531fe426/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L1333l) directly assigns patch value without checking it's inners.

Thus, it causes;

```
Before: {}
Patch: {"complex":{"nested":null}}
After: {"complex":{"nested":null}}
According to the PR correct result: {}
```
or
```
Before: {}
Patch: {"complex":{"nested":null,"nested2":"foo"}}
After: {"complex":{"nested":null,"nested2":"foo"}}
According to the PR correct result: {"complex":{"nested2":"foo"}}
```

(note: I added above tests as unit test in PR)

When I check to assure that is compatible with jsonmerge, I added this test in [here](https://github.com/kubernetes/kubernetes/blob/876d4e0ab029ac7c314bb0e033bdd036531fe426/staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go#L229);

```
    original:
      simpleMap:
        key1: 1
    modified:
      simpleMap:
        key1: 1
        key2:
         key3: null
    current:
      simpleMap:
        key1: a
        other: a
    threeWay:
      simpleMap:
        key1: 1
        key2:
          key3: null
    result:
      simpleMap:
        key1: 1
        key2: {}
        other: a
```

It also removes `key3`, the only difference is it added `key2: {}`, but this PR removes also it.

apelisse replied:
Yeah, thanks for that, do you know if we can solve that here? i.e. preserve `key2`.

In that test, we would have:
```
ExpectedTwoWayResult: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{},"name":"foo","scalar":true}`
```

ardaguclu replied:
@apelisse, thanks for reviews. 

I changed the PR to preserve  `complex_nullable` and also the unit tests according to it.

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -6713,6 +6713,36 @@ func TestUnknownField(t *testing.T) {
+		"discard empty slices": {
+			Original: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Current:  `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Modified: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":[],"name":"foo","scalar":true}`,
```
empty slice is not the same as null slice, we should keep these.

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -6713,6 +6713,36 @@ func TestUnknownField(t *testing.T) {
+			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			ExpectedThreeWay:       `{"complex_nullable":[]}`,
+			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+		},
```
Do we have a test that shows that we keep null values in the object?

i.e.:
```
Current: `{"something": null}`
ExpectedTwoWay: `{}`
ExpectedTwoWayResult: `{"something": null}
```

ardaguclu replied:
I tried this preserve test in master branch;

```
"added only": {
			Original: `{"name":"foo"}`,
			Current:  `{"name":"foo","exist":null}`,
			Modified: `{"name":"foo","scalar":true,"complex":{"nested":true},"array":[1,2,3]}`,

			ExpectedTwoWay:         `{"array":[1,2,3],"complex":{"nested":true},"scalar":true}`,
			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"exist":null,"name":"foo","scalar":true}`,
			ExpectedThreeWay:       `{"array":[1,2,3],"complex":{"nested":true},"scalar":true}`,
			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"exist":null,"name":"foo","scalar":true}`,
		},
```

It failed with;

```
using strategicpatch.PatchMetaFromStruct in testcase added only and preserve nulls: expected two-way result:
        	{"array":[1,2,3],"complex":{"nested":true},"exist":null,"name":"foo","scalar":true}
        got
        	{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}
```

AFAICS, strategic patch removes all null fields during patch from current. This PR aligns this behavior applying also to modified.

Isn't this the correct behavior?, if not maybe this patch should focus on preserving null fields to preserve idempotency?

apelisse replied:
No, the object should be able to carry null values without having these removed by the patch. Null values *in the patch* instruct that the fields should be removed.

If you have an object with `{"a": null, "b": "some value"}`, why would a patch that changes `b` (e.g. `{"b": "new value"}` also remove `a`?

ardaguclu replied:
@apelisse I definitely agree with you about in the example you gave ^ `a` should not be removed.

However, I checked jsonmergepatch as reference;

```
    original:
      simpleMap:
        key1: 1
        key2: 1
    modified:
      simpleMap:
        key1: 1
    current:
      simpleMap:
        key1: 1
        key2: 1
        other: a
    threeWay:
      simpleMap:
        key2: null
    result:
      simpleMap:
        key1: 1
        other: a
```

`key2` is not in modified but it is removed.

apelisse replied:
Side note/Disclaimer/Rant
---
Yeah, these test confuse me because of the "3 way" vs "2 way" and "modified" vs "current". They don't help test what we're trying to test to be honest.

I guess in most of these tests, I mostly look at "current" "twoway" and "twowayresult" as a way to see how the patch is applied (the 3 way aspects are mostly related to the client AFAIU).

Actual answer :-)
---
So I merely look at "current", "threeway" (the patch), and "result", I see:
The map has 3 fields, the patch asks key2 to be removed, the map has 2 fields with the field remove. Exactly what I would expect. Not sure that's a problem.

Here's something I'd like to see:
```
    original:
      simpleMap:
        key1: 1
    modified:
      simpleMap:
    current:
      simpleMap:
        key1: 1
    threeWay:
      simpleMap:
        key1: null # is that correct?
    result:
      simpleMap:
```

ardaguclu replied:
Yes, I think your expected result is similar to my findings and I think there is no problem at that part. 

But the problem related to the real issue is in above thread and this PR fixes it. In this thread, mostly we were focusing on "is there any other problem" :).

ardaguclu replied:
Why I'm saying this because if we filter the nested null values from patchV as this PR suggested. There will no more any problem.

apelisse reviewed (commented):
Thanks!

MadhavJivrajani commented:
/ok-to-test

ardaguclu commented:
This looks a unrelated transient failure;

/test pull-kubernetes-unit

fedebongio commented:
/assign @apelisse 
/cc @jpbetz 
/triage accepted

ardaguclu commented:
/test pull-kubernetes-e2e-kind-ipv6

apelisse reviewed (commented):
Thanks, can we also have an additional test to make sure that nulls are not removed from the existing object, probably a test like this:
```
"null values are preserved": {
	Original: `{"thing": null, "nested": {"thing": null}}`,
	Current:  `{"thing": null, "nested": {"thing": null}}`,
	Modified: `{"thing": null, "nested": {"thing": null, "value": 5}}`,

	ExpectedTwoWay:         `{"nested":{"value":5}}`,
	ExpectedTwoWayResult:   `{"thing": null, "nested": {"thing": null, "value": 5}}`,
	ExpectedThreeWay:       `{"nested":{"value":5}}`,
	ExpectedThreeWayResult: `{"thing": null, "nested": {"thing": null, "value": 5}}`,
},
```

apelisse commented:
Looking at the code, this is clearly a bug (unanticipated use-case by original author). I think the code looks good and the tests look good, thanks!

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1375,6 +1383,30 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
+		return nil
+	}
+
+	patchT := reflect.TypeOf(patchV)
+	switch patchT.Kind() {
+	case reflect.Map:
+		filteredPatch := make(map[string]interface{})
+		for k, v := range patchV.(map[string]interface{}) {
+			if f := discardNullValuesFromPatch(v); f != nil {
+				filteredPatch[k] = f
+			}
+		}
+
+		return filteredPatch
+
+	default:
+		return patchV
+	}
```
A few observations:
1. I'm not sure what the reflection is getting us that a direct type cast attempt wouldn't do
2. Kind() == Map doesn't guarantee it is a `map[string]interface{}`, right?
3. The only calls here are from a function that says it can modify the patch... should we delete null elements from the patch instead of making a copy of each level?
4. The godoc says it recurses into slices, but I don't see that happening... should it be? if so, we need a test for that as well

I wonder if this could be simplified to something like:
```go
switch patchV := patchV.(type) {
case map[string]interface{}:
  for k, v := range patchV {
    if v == nil {
      delete(patchV, k)
    } else {
      discardNullValuesFromPatch(v)
    }
  }

case []interface{}:
  for _, v := range patchV {
    discardNullValuesFromPatch(v)
  }

}

return patchV
```

ardaguclu replied:
> A few observations:
> 
> 1. I'm not sure what the reflection is getting us that a direct type cast attempt wouldn't do
> 2. Kind() == Map doesn't guarantee it is a `map[string]interface{}`, right?

I'm definitely in favor of using type casting instead reflection. But as far as I can see, there are lots of places using reflection in this package, and I thought that reflection is being used due to a **_reason_**.
> 3. The only calls here are from a function that says it can modify the patch... should we delete null elements from the patch instead of making a copy of each level?

Good idea
> 4. The godoc says it recurses into slices, but I don't see that happening... should it be? if so, we need a test for that as well
> 

First attempt tried to also iterate over slices but then I removed it. Maybe we should put it back. What do you think @apelisse ?
> I wonder if this could be simplified to something like:
> 
> ```go
> switch patchV := patchV.(type) {
> case map[string]interface{}:
>   for k, v := range patchV {
>     if v == nil {
>       delete(patchV, k)
>     } else {
>       discardNullValuesFromPatch(v)
>     }
>   }
> 
> case []interface{}:
>   for _, v := range patchV {
>     discardNullValuesFromPatch(v)
>   }
> 
> }
> 
> return patchV
> ```

That function ^ looks good to me.

apelisse commented:
Yeah, we should recurse in associative lists (but probably not in non-associative lists to be honest), but yeah, what I think I said is that setting a list to empty shouldn't remove the list.

ardaguclu commented:
@apelisse @liggitt I changed the patch to also support associative slices(and of course, added tests for that). Could you please take a look at?, thanks.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1375,6 +1383,29 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
 	return original, nil
 }
 
+// discardNullValuesFromPatch discards all null values from patch.
+// It traverses for all slices and map types to discard all nulls.
```
clarify this discards null property values, not null list items

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1375,6 +1383,29 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
+	case []interface{}:
+		patchS := patchV.([]interface{})
+		for i, v := range patchV.([]interface{}) {
+			patchS[i] = discardNullValuesFromPatch(v)
```
I don't think this needs to reassign to `patchS[i]`:
```suggestion
			discardNullValuesFromPatch(v)
```

ardaguclu commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1375,6 +1383,29 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
+// discardNullValuesFromPatch discards all null values from patch.
+// It traverses for all slices and map types to discard all nulls.
+func discardNullValuesFromPatch(patchV interface{}) interface{} {
+	switch patchV.(type) {
```
I did not know that feature in Go, that really simplified. Thanks.

ardaguclu committed:
Simplify casting in discardNullValuesFromPatch

ardaguclu commented:
/test pull-kubernetes-unit

liggitt commented:
/lgtm
/approve

pacoxu commented:
/test pull-kubernetes-integration
for timeout
